### ISSUE TITLE
Incompatibility with Ron crate prevents Nutype deserialization from RON format

### ISSUE THREAD

**Issue by OmegaJak** [greyblake/nutype#87]
The [Rusty Object Notation](https://github.com/ron-rs/ron) crate provides a data serialization format. When a String-based Nutype that derives Serialize and Deserialize is serialized to RON, it cannot be deserialized back from RON. It produces an `ExpectedString` error during deserialization. I believe this may be due to the implementation of the `Deserialize` derive in Nutype. The following two tests demonstrate the issue.

```rust
#[cfg(feature = "serde")]
#[test]
fn test_ron_roundtrip_with_nutype() {
    #[nutype::nutype(derive(PartialEq, Debug, Serialize, Deserialize))]
    struct NuTypeString(String);

    let value = NuTypeString::new("foo");

    let serialized = ron::ser::to_string(&value).unwrap();
    println!("Serialized as:{}", serialized);
    let deserialized: Result<NuTypeString, _> = ron::de::from_str(&serialized);
    
    assert_eq!(Ok(value), deserialized);
}

#[cfg(feature = "serde")]
#[test]
fn test_ron_roundtrip_without_nutype() {
    #[derive(PartialEq, Debug, serde::Serialize, serde::Deserialize)]
    struct SerdeNewTypeString(String);

    let value = SerdeNewTypeString("foo".to_string());

    let serialized = serde_json::to_string(&value).unwrap();
    println!("Serialized as:{}", serialized);
    let deserialized: SerdeNewTypeString = serde_json::from_str(&serialized).unwrap();
    
    assert_eq!(value, deserialized);
}
```

When executed on current master (commit 748ce299), with Ron v0.8.1 and the serde "derive" feature enabled, `test_ron_roundtrip_with_nutype` fails and `test_ron_roundtrip_without_nutype` passes.

This is the output from `test_ron_roundtrip_with_nutype`:
```
Serialized as:("foo")
thread 'derives::test_ron_roundtrip_with_nutype' panicked at 'assertion failed: `(left == right)`
  left: `Ok(NuTypeString("foo"))`,
 right: `Err(SpannedError { code: ExpectedString, position: Position { line: 1, col: 1 } })`', test_suite\tests\string.rs:500:9
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library\std\src\panicking.rs:593
   1: core::panicking::panic_fmt
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library\core\src\panicking.rs:67
   2: core::fmt::Arguments::new_v1
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library\core\src\fmt\mod.rs:311
   3: core::panicking::assert_failed_inner
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library\core\src\panicking.rs:274
   4: core::panicking::assert_failed<enum2$<core::result::Result<string::derives::test_ron_roundtrip_with_nutype::__nutype_private_NuTypeString__::NuTypeString,ron::error::SpannedError> >,enum2$<core::result::Result<string::derives::test_ron_roundtrip_with_nuty
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26\library\core\src\panicking.rs:228
   5: string::derives::test_ron_roundtrip_with_nutype
             at .\tests\string.rs:500
   6: string::derives::test_ron_roundtrip_with_nutype::closure$0
             at .\tests\string.rs:490
   7: core::ops::function::FnOnce::call_once<string::derives::test_ron_roundtrip_with_nutype::closure_env$0,tuple$<> >    
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26\library\core\src\ops\function.rs:250
   8: core::ops::function::FnOnce::call_once
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library\core\src\ops\function.rs:250
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test derives::test_ron_roundtrip_with_nutype ... FAILED
```

I believe this is because Ron wraps the serialized value in parentheses, while Nutype's derive implementation attemps to directly deserialize the whole thing as a string, expecing quotes to come first.

Comparing the generated code, we can see the differences between Nutype's `Deserialize` implementation and the one generated by serde:
<details>
<summary>Click to expand code</summary>

```rust
impl<'de> ::serde::Deserialize<'de> for NuTypeString {
    fn deserialize<D: ::serde::Deserializer<'de>>(
        deserializer: D,
    ) -> Result<Self, D::Error> {
        let raw_value = String::deserialize(deserializer)?;
        Ok(NuTypeString::new(raw_value))
    }
}
// --- Nutype ↑ VS Serde ↓--- //
impl<'de> _serde::Deserialize<'de> for SerdeNewTypeString {
    fn deserialize<__D>(
        __deserializer: __D,
    ) -> _serde::__private::Result<Self, __D::Error>
    where
        __D: _serde::Deserializer<'de>,
    {
        #[doc(hidden)]
        struct __Visitor<'de> {
            marker: _serde::__private::PhantomData<SerdeNewTypeString>,
            lifetime: _serde::__private::PhantomData<&'de ()>,
        }
        impl<'de> _serde::de::Visitor<'de> for __Visitor<'de> {
            type Value = SerdeNewTypeString;
            fn expecting(
                &self,
                __formatter: &mut _serde::__private::Formatter,
            ) -> _serde::__private::fmt::Result {
                _serde::__private::Formatter::write_str(
                    __formatter,
                    "tuple struct SerdeNewTypeString",
                )
            }
            #[inline]
            fn visit_newtype_struct<__E>(
                self,
                __e: __E,
            ) -> _serde::__private::Result<Self::Value, __E::Error>
            where
                __E: _serde::Deserializer<'de>,
            {
                let __field0: String = match <String as _serde::Deserialize>::deserialize(
                    __e,
                ) {
                    _serde::__private::Ok(__val) => __val,
                    _serde::__private::Err(__err) => {
                        return _serde::__private::Err(__err);
                    }
                };
                _serde::__private::Ok(SerdeNewTypeString(__field0))
            }
            #[inline]
            fn visit_seq<__A>(
                self,
                mut __seq: __A,
            ) -> _serde::__private::Result<Self::Value, __A::Error>
            where
                __A: _serde::de::SeqAccess<'de>,
            {
                let __field0 = match match _serde::de::SeqAccess::next_element::<
                    String,
                >(&mut __seq) {
                    _serde::__private::Ok(__val) => __val,
                    _serde::__private::Err(__err) => {
                        return _serde::__private::Err(__err);
                    }
                } {
                    _serde::__private::Some(__value) => __value,
                    _serde::__private::None => {
                        return _serde::__private::Err(
                            _serde::de::Error::invalid_length(
                                0usize,
                                &"tuple struct SerdeNewTypeString with 1 element",
                            ),
                        );
                    }
                };
                _serde::__private::Ok(SerdeNewTypeString(__field0))
            }
        }
        _serde::Deserializer::deserialize_newtype_struct(
            __deserializer,
            "SerdeNewTypeString",
            __Visitor {
                marker: _serde::__private::PhantomData::<SerdeNewTypeString>,
                lifetime: _serde::__private::PhantomData,
            },
        )
    }
}
```
</details>

I'm not certain exactly what serde's impl is doing, but somehow it handles the parentheses at the start, where Nutype's doesn't. This prevented me from fully using a Nutype-based struct as persistent data while using egui, which uses RON to persist data on disk. I instead ended up manually converting to/from a string when interacting with my persistence struct, losing some of the benefits Nutype provides.

**Comment by greyblake** [comment]
@OmegaJak Thanks for reporting this and in particular for the detailed report.

I admit, I haven't test Serialize/Deserialize with other libraries except serde_json.

I'll try to investigate it and we what I can do. Hopefully there is some shortcut.

**Comment by greyblake** [comment]
@OmegaJak Btw, your second test is using serde_json, not ron:
![image](https://github.com/greyblake/nutype/assets/113512/985d4951-9ddd-4636-8a0e-d1430909b015)

But I assume it works with ron the same way.

**Comment by OmegaJak** [comment]
@greyblake Whoops yeah sorry about that. Updated test still works with ron:
```rust
#[cfg(feature = "serde")]
#[test]
fn test_ron_roundtrip_without_nutype() {
    #[derive(PartialEq, Debug, serde::Serialize, serde::Deserialize)]
    struct SerdeNewTypeString(String);

    let value = SerdeNewTypeString("foo".to_string());

    let serialized = ron::ser::to_string(&value).unwrap();
    println!("Serialized as:{}", serialized);
    let deserialized: Result<SerdeNewTypeString, _> = ron::de::from_str(&serialized);
    
    assert_eq!(Ok(value), deserialized);
}
```

**Comment by greyblake** [comment]
@OmegaJak This is fixed in https://github.com/greyblake/nutype/pull/89 and merged to master.

If it's possible for you, you may consider using the version from `master` for a while.
If you urgently need a release with this, fix then I may consider backporting in to 0.3.x and publishing a new version.

**Comment by greyblake** [comment]
The functionality is available in `0.4.0-beta.1` at the moment:
https://crates.io/crates/nutype/0.4.0-beta.1

**Comment by OmegaJak** [comment]
@greyblake Great thanks, I'll check this out today and report back

**Comment by OmegaJak** [comment]
@greyblake Pointing to git master works great, version v0.4.0-beta.1 doesn't - it has the same issue as before. This appears to be because the `Cargo.toml` of the main `nutype` package is still pointing at `nutype_macros` v0.3.1 instead of v0.4.0-beta.1. My project is fine running off master for a time anyway, though.

Thanks for fixing this so quickly!

**Comment by greyblake** [comment]
@OmegaJak Oh gosh, shame on me. Thanks for the report, I am glad it was only beta.
I just published `v0.4.0-beta.2`  with proper versions =)

**Comment by OmegaJak** [comment]
@greyblake Yep, all good now in `0.4.0-beta.2`. That's what betas are for 😉
